examples/cis2-dynamic-nft/src/lib.rs

```diff
@@ -0,0 +1,1450 @@
+//! A dynamic token example implementation of the Concordium Token Standard
+//! CIS2.
+//!
+//! # Description
+//! An instance of this smart contract can contain a number of different token
+//! types each identified by a token ID. A token type is then globally
+//! identified by the contract address together with the token ID.
+//!
+//! In this example the contract is initialized with no tokens, and tokens can
+//! be minted through a `mint` contract function, which will only succeed for
+//! the contract owner. No functionality to burn token is defined in this
+//! example.
+//!
+//! `mint` function takes a list of urls for the metadata, instead of one url.
+//! contract owner has a right to change the token by calling `upgrade`
+//! function. or contract owner can add a new metadata as the last url of a
+//! list. Intentionally remove metadata is not implemented.
+//!
+//! `tokenMetadataList` function shows all metadata history.
+//! `addMetadata`function adds new metadata.
+//!
+//! Note: The word 'address' refers to either an account address or a
+//! contract address.
+//!
+//! As follows from the CIS2 specification, the contract has a `transfer`
+//! function for transferring an amount of a specific token type from one
+//! address to another address. An address can enable and disable one or more
+//! addresses as operators. An operator of some address is allowed to transfer
+//! any tokens owned by this address.
+//!
+//! This contract also contains an example of a function to be called when
+//! receiving tokens. In which case the contract will forward the tokens to
+//! the contract owner.
+//! This function is not very useful and is only there to showcase a simple
+//! implementation of a token receive hook.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use concordium_cis2::*;
+use concordium_std::*;
+
+/// List of supported standards by this contract address.
+const SUPPORTS_STANDARDS: [StandardIdentifier<'static>; 2] =
+    [CIS0_STANDARD_IDENTIFIER, CIS2_STANDARD_IDENTIFIER];
+
+/// Custom Event
+pub const TOKEN_UPGRADE_EVENT_TAG: u8 = 0u8;
+
+/// The RegistrationEvent is logged when a new public key is registered.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct TokenEvolvedEvent {
+    /// Account that a public key will be registered to.
+    token_id: ContractTokenId,
```

Comment:
It is nice to also emit the new metadata url in my opinion.

Reply:
added.